src/compiler/crystal/syntax/parser.cr

```diff
@@ -2083,15 +2085,22 @@ module Crystal
         line = remove_heredoc_from_line(line, indent, pieces.last.line_number) if remove_indent
         add_heredoc_piece new_pieces, line
       end
-      new_pieces
+      new_pieces.map do |piece|
```

Comment:
Can this `map` be avoided and instead this logic put in `add_heredoc_piece` somehow?

Reply:
To avoid this `map`, `add_heredoc_piece` must take extra argument `last_is_string` and it must return new `last_is_string` state. Like:

```crystal
    def remove_heredoc_indent(pieces : Array, indent)
      ...
      last_is_string = false
      new_pieces = [] of ASTNode
      ...

      pieces.each_with_index do |piece, index|
        ...
        last_is_string = add_heredoc_piece(pieces, piece, last_is_string)
        ...
      end

      ...
    end

    private def add_heredoc_piece(pieces, piece : String, last_is_string : Bool)
      if last_is_string
        pieces.last!.as(StringLiteral).value += piece
      else
        pieces << StringLiteral.new(piece)
      end
      true
    end

    private def add_heredoc_piece(pieces, piece : ASTNode, last_is_string : Bool)
      pieces << piece
      false
    end
``